library/core/src/option.rs

```diff
@@ -1869,6 +1869,121 @@ impl<T, E> Option<Result<T, E>> {
     }
 }
 
+/// Implements the various `Fn` traits to make the `transpose_fn` family of
+/// methods work over any functions.
+///
+/// This is not otherwise possible with a closure due to `|args: A|` being
+/// treated as `Fn<(A,)>` instead of `Fn<A>`.
+struct TransposeFn<F> {
+    f: Option<F>,
+}
+
+impl<F> Option<F> {
+    /// Transposes an `Option` of an [`Fn`] into an [`Fn`] that returns an
+    /// `Option`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(transpose_fn)]
+    ///
+    /// let x = Some(|| 42);
+    /// let y = x.transpose_fn();
+    ///
+    /// assert_eq!(y(), Some(42));
+    /// ```
+    #[inline]
+    #[unstable(feature = "transpose_fn", issue = "none")]
+    pub fn transpose_fn<A>(self) -> impl Fn<A, Output = Option<F::Output>>
```

Comment:
A limitation of using `impl Trait` rather than exposing the concrete return type publicly is that the only conditional traits the return type has are `Send` & `Sync`. If this used a concrete return type, it could also conditionally implement `Clone` & `Copy`.

However, I think that for an MVP of this feature is fine with `impl Trait` since this can be changed in a later commit.

Reply:
On second thought, if the `TransposeFn` type is exposed, then this API could just be a single function rather than three. So perhaps it should be part of the MVP.